src/cc/BPF.cc

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdio>
+#include <exception>
+#include <linux/bpf.h>
+#include <memory>
+#include <unistd.h>
+#include <vector>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "libbpf.h"
+#include "perf_reader.h"
+
+#include "BPF.h"
+
+namespace ebpf {
+
+  std::string uint_to_hex(uint64_t value) {
+    std::stringstream ss;
+    ss << std::hex << value;
+    return ss.str();
+  }
+
+  std::string sanitize_str(
+    std::string str, bool (*validator)(char), char replacement = '_'
+  ) {
+    for (size_t i = 0; i < str.length(); i++)
+      if (!validator(str[i]))
+        str[i] = replacement;
+    return str;
+  }
+
+  BPF::BPF(
+    const std::string& bpf_program,
+    unsigned int debug_flag, const char* cflags[], int ncflags
+  ) :
+    bpf_module_(new BPFModule(debug_flag))
+  {
+    if (bpf_module_->load_string(bpf_program, cflags, ncflags) != 0)
+      throw std::runtime_error("Unable to initialize BPF program");
```

Comment:
would be nice for there to be a map of error numbers to messages somewhere and to throw to msg in with the exception

bonus points if there's a BPFException class